python/ray/dataframe/dataframe.py

```diff
@@ -555,16 +621,188 @@ def expanding(self, min_periods=1, freq=None, center=False, axis=0):
         raise NotImplementedError("Not Yet implemented.")
 
     def ffill(self, axis=None, inplace=False, limit=None, downcast=None):
-        raise NotImplementedError("Not Yet implemented.")
+        new_df = self.fillna(
+            method='ffill', axis=axis, limit=limit, downcast=downcast
+        )
+        if inplace:
+            self._df = new_df._df
+            self.columns = new_df.columns
+        else:
+            return new_df
 
     def fillna(self, value=None, method=None, axis=None, inplace=False,
                limit=None, downcast=None, **kwargs):
-        raise NotImplementedError("Not Yet implemented.")
+        if isinstance(value, (list, tuple)):
+            raise TypeError('"value" parameter must be a scalar or dict, but '
+                            'you passed a "{0}"'.format(type(value).__name__))
+        if value is None and method is None:
+            raise ValueError('must specify a fill method or value')
+        if value is not None and method is not None:
+            raise ValueError('cannot specify both a fill method and value')
+        if method is not None and method not in ['backfill', 'bfill', 'pad',
+                                                 'ffill']:
+            expecting = 'pad (ffill) or backfill (bfill)'
+            msg = 'Invalid fill method. Expecting {expecting}. Got {method}'\
+                  .format(expecting=expecting, method=method)
+            raise ValueError(msg)
+
+        new_df = self._map_partitions(
+            lambda df: df.fillna(value=value, method=method, axis=axis,
+                                 limit=limit, downcast=downcast, **kwargs)
+        )
+        is_bfill = method is not None and method in ['backfill', 'bfill']
+        is_ffill = method is not None and method in ['pad', 'ffill']
+        is_axis_zero = axis is None or axis == 0
+
+        if is_axis_zero:
+            if is_bfill or is_ffill:
```

Comment:
I would prefer this not be nested like this. I think it would be more readable as:

```python
if is_axis_zero and (is_bfill or is_ffill):
    ...
else:
    ...
```